### DISCUSSION TITLE
Hidden field visibility across package boundaries leads to unexpected output&behaviour

### DISCUSSION THREAD

**Discussion by jpluscplusm** [cue-lang/cue#2205] [Q&A]
Hi all 👋 

This testscript passes, doing in-package "functional programming" with CUE:

```
exec cue eval .:p
stdout 'out: 3'

-- consumer.cue --
package p

fpc: #Sum: {
	_x:  int
	_y:  int
	out: _x + _y
}

test: fpc.#Sum & {
	_x: 1
	_y: 2
}
```

This testscript fails, as the "function" has been moved outside the current package and the hidden-fields-as-params concept breaks down:

```
exec cue eval .:p
stdout 'out: 3'

-- consumer.cue --
package p

import "cue.test/fpc"

test: fpc.#Sum & {
	_x: 1
	_y: 2
}

-- cue.mod/module.cue --
module: ""

-- cue.mod/pkg/cue.test/fpc/sum.cue --
package fpc

#Sum: {
	_x:  int
	_y:  int
	out: _x + _y
}
```

Question 1 for y'all: is the FP-in-CUE pattern simply uninteresting enough that it /not/ working when "functions" are moved into their own package ... is fine?

[Edit: I understand that FP works when *non*-hidden fields are used for passing params in - but that leads to the "return" type of a "function" carrying its input params downstream, to all consumers of the function. And I /believe/ it also rules out the function-emitting-a-single-CUE-value-via-an-embedding pattern; which, while out of scope for this discussion, is still an incredibly useful technique worth keeping in mind here, at a minimum. I'll open another discussion around it, sometime soon.]

Here's why I ask Q1, above: the `eval` in the above, failing, testscript demonstrably *does* have enough information that it should "work":

```
exec cue eval --show-hidden .:p
cmp stdout expected.output
exec cue export .:p
stdout 'out: 3'

-- consumer.cue --
package p

import "cue.test/fpc"

test: fpc.#Sum & {
	_x: 1
	_y: 2
}

-- cue.mod/module.cue --
module: ""

-- cue.mod/pkg/cue.test/fpc/sum.cue --
package fpc

#Sum: {
	_x:  int
	_y:  int
	out: _x + _y
}

-- expected.output --
test: {
    _x:  1
    out: _x + _y
    _y:  2
}
```

Here, `expected.output` clearly contains "enough" info that it should be able to emit a concrete value for `out`. The fact that the  following testscript *does* work demonstrates as much:

```
exec cue eval --show-hidden .:p
cmp stdout expected.cue.eval.output
exec bash -c 'cue eval --show-hidden .:p | cue export -'
cmp stdout expected.cue.eval.piped.into.cue.export.output

-- consumer.cue --
package p

import "cue.test/fpc"

test: fpc.#Sum & {
	_x: 1
	_y: 2
}

-- cue.mod/module.cue --
module: ""

-- cue.mod/pkg/cue.test/fpc/sum.cue --
package fpc

#Sum: {
	_x:  int
	_y:  int
	out: _x + _y
}

-- expected.cue.eval.output --
test: {
    _x:  1
    out: _x + _y
    _y:  2
}
-- expected.cue.eval.piped.into.cue.export.output --
{
    "test": {
        "out": 3
    }
}
```

Question 2: what gives, here?

(I've tested all these examples with both 0.4.3 and 0.5.0 beta 2.)

IMHO:

- hidden fields should be permitted to be used as cross-package "params", with whatever minimal visibility and/or evaluation-order changes that demands. What do you reckon?

- if they're not, then /something/ should change such that the eval-piped-into-export example doesn't work. It's a whiffy hack that'd be horrible to have to suggest as a workaround, if it ends up working long-term! (IMHO!)

**Comment by verdverm** [comment]
hidden fields are not visible across package boundaries, this provides a "private" space for package developers

use definitions for input parameters if you do not want them as part of the output

```cue
#Sum: {
	#x:  int
	#y:  int
	out: #x + #y
}

z: (#Sum & { #x: 1, #y: 2}).out
```
